### PR TITLE
feat: Add customization for Windows guest OS

### LINKED ISSUE #177: `vsphere-clone`: Add support for OS customization for Windows
Working on Windows template creation, I noticed that the OS customizations for windows are not supported in packer: there is a very suitable `linux_options` for Linux but no `windows_options`.

A number of issues have been raised related that would be solved by adding such support:
- https://github.com/hashicorp/packer-plugin-vsphere/issues/7 (workaround is not able to customize the network settings)
- Setting a static IP address requires crafting an unattend.xml file: https://github.com/jetbrains-infra/packer-builder-vsphere/issues/37

I tried porting the OS customization for Windows from the terraform implementation to packer on a personal fork of the plugin and was successful. This is the code I ported: https://github.com/hashicorp/terraform-provider-vsphere/blob/773942213dd548ac5e864dc62816e97442cdda57/vsphere/internal/vmworkflow/virtual_machine_customize_subresource.go#L107 and its documentation: https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine (then search for `windows_options`)
We had some initial discussion in #174  and decided it was worth discussing this in its own issue.

Questions we raised:
- using the terraform semantics ? For example as opposed to govc that requires to point at a static customization profile by name and then layering changes on the top - I tried that and got lost by the way.
- expose all of the OS customization including the ones that should probably be avoided for template baking or keep some of them out? For example AD Join is debatable: #34.

Please do edit this issue (or make a new issue) if this framing for the feature request is not helpful.

### PR DESCRIPTION
**Pull request by hmalphettes**
Ported from terraform vsphere virtual machine resource

Closes #177

### CODE REVIEW

File: docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx [hunk 1]
```diff
@@ -1,11 +1,13 @@
-<!-- Code generated from the comments of the CustomizeConfig struct in builder/vsphere/clone/step_customize.go; DO NOT EDIT MANUALLY -->
+<!-- Code generated from the comments of the CustomizeConfig struct in builder\vsphere\clone\step_customize.go; DO NOT EDIT MANUALLY -->
```
Comment (hmalphettes) [inline docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx:1]: looks like packer-sdc needs to generate forward slashes even when executed on windows.

**Comment by hmalphettes** [comment]
To test this pull request, compile the packer plugin as `packer-plugin-vspherewinopts` and use some example like:

```hcl
source "vspherewinopts-clone" "testit" {
  ...
  communicator        = "none"
  disable_shutdown   = true
  shutdown_timeout  = "30m"

  customize {
    windows_options {
      computer_name  = "bakingwindowstest"
      admin_password = "s3cr3t"
      organization_name = "yourorg"
      auto_logon = true
      auto_logon_count = 1
      run_once_command_list = [
         "C:\\Windows\\System32\\WindowsPowershell\\v1.0\\powershell.exe -Command \"Enable-NetAdapterBinding -Name Ethernet0 -ComponentID ms_tcpip\""
      ]
      time_zone = 215
    }

    network_interface {
      ipv4_address = var.vm_ip
      ipv4_netmask = var.vm.netmask_cidr
      dns_server_list = [var.vm.dns_server]
    }
    ipv4_gateway = local.network_gateway
  }
}
```

And use the web console to check the content of `panther/unattend.xml`

File: docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx (hunk 1 above)
Comment (hmalphettes) [reply docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx]: Issue with packer-sdc raised here: https://github.com/hashicorp/packer-plugin-sdk/issues/104
Note: for this pull requerst, I have manually edited the comments as a workaround until this issue is solved.

File: builder/vsphere/clone/step_customize.go [hunk 2]
```diff
@@ -49,6 +51,39 @@ type LinuxOptions struct {
...
+	AdminPassword *string `mapstructure:"admin_password"`
+	// The new time zone for the virtual machine. This is a sysprep-dictated timezone code. Default 85 (GMT)
+	TimeZone *int32 `mapstructure:"time_zone"`
+	// CustomizationIdentification
+	// The user account of the domain administrator used to join this virtual machine to the domain.
+	DomainAdminUser string `mapstructure:"domain_admin_user"`
+	// The password of the domain administrator used to join this virtual machine to the domain.
+	DomainAdminPassword string `mapstructure:"domain_admin_password"`
+	// The domain that the virtual machine should join.
+	JoinDomain string `mapstructure:"join_domain"`
```
Comment (tenthirtyam) [inline builder/vsphere/clone/step_customize.go:73]: I don't think it makes good sense to include the ability to join the machine image to a domain per comments in #34.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

File: builder/vsphere/clone/step_customize.go [hunk 3]
```diff
@@ -273,4 +302,135 @@ func (s *StepCustomize) globalIpSettings() types.CustomizationGlobalIPSettings {
...
+		}
+	}
+	if w.JoinDomain != "" {
+		if w.DomainAdminUser == "" {
+			errs = append(errs, fmt.Errorf("The `domain_admin_user` is required when `join_domain` is specified"))
+		}
+		if w.DomainAdminPassword == "" {
+			errs = append(errs, fmt.Errorf("The `domain_admin_password` is required when `join_domain` is specified"))
+		}
+	}
```
Comment (tenthirtyam) [inline builder/vsphere/clone/step_customize.go:350]: I don't think it makes good sense to include the ability to join the machine image to a domain per comments in #34.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

File: builder/vsphere/clone/step_customize.go [hunk 4]
```diff
@@ -273,4 +302,135 @@ func (s *StepCustomize) globalIpSettings() types.CustomizationGlobalIPSettings {
...
+		DomainAdmin:   w.DomainAdminUser,
+	}
+	if w.AdminPassword != nil {
+		obj.DomainAdminPassword = &types.CustomizationPassword{
+			Value:     *w.AdminPassword,
+			PlainText: true,
+		}
+	}
+	return obj
+}
```
Comment (tenthirtyam) [inline builder/vsphere/clone/step_customize.go:422]: I don't think it makes good sense to include the ability to join the machine image to a domain per comments in #34.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+- `admin_password` (\*string) - The new administrator password for this virtual machine.
+
+- `time_zone` (\*int32) - The new time zone for the virtual machine. This is a sysprep-dictated timezone code. Default 85 (GMT)
+
+- `domain_admin_user` (string) - CustomizationIdentification
+  The user account of the domain administrator used to join this virtual machine to the domain.
+
+- `domain_admin_password` (string) - The password of the domain administrator used to join this virtual machine to the domain.
+
+- `join_domain` (string) - The domain that the virtual machine should join.
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:20]: I don't think it makes good sense to include the ability to join the machine image to a domain per comments in #34.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Review by tenthirtyam — COMMENTED** [review]
The title and description of the PR is not enough and is misleading. Additional options for Linux are also included.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

File: docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx (hunk 1 above)
Comment (hmalphettes) [reply docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx]: The fix for packer-sdc on windows was merged.

**Comment by nywilken** [comment]
Hi @hmalphettes thanks for contribution. It looks like there are a few requested changes and items in progress. I'm going to convert this PR into a draft now. When ready please give @hashicorp/packer and we will gladly review the changes. Cheers!

Apologies for the delayed review. Many thanks to @tenthirtyam for helping push the changes forward.

**Comment by hmalphettes** [comment]
@nywilken, no problem at all. Give me 1 or 2 days to rebase, remove the AD join and cleanup this PR.

File: builder/vsphere/clone/step_customize.go (hunk 3 above)
Comment (hmalphettes) [reply builder/vsphere/clone/step_customize.go]: ok removed.

File: builder/vsphere/clone/step_customize.go (hunk 2 above)
Comment (hmalphettes) [reply builder/vsphere/clone/step_customize.go]: OK.

**Comment by hmalphettes** [comment]
@hashicorp/packer I rebased and removed the AD Join related parameters.

As mentioned by @tenthirtyam I did refactor the parameters validation for linux to mirror the same code for windows.
If you would rather have that in a separate pull request, let me know and I'll revert that back.

Thanks in advance for your kind review.

On a practical basis, let's highlight that using DHCP during the bake of a Windows template makes everything simpler:
in particular: the packer config are simpler and the scripts in the guest OS too.
Without extra care the VMs will boot with the static IP it used during the packer build.
Being able to use packer with a static IP address can unblock situations where there is no DHCP server.
But using DHCP during a packer build I think is a best practice.

**Comment by nywilken** [comment]
Hi @hmalphettes thanks for making the updates. Taking this PR out of draft so that we can give it a review. Cheers!

File: docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx [hunk 6]
```diff
@@ -2,6 +2,8 @@
 
 - `linux_options` (\*LinuxOptions) - Settings to Linux guest OS customization. See [Linux customization settings](#linux-customization-settings).
 
+- `windows_options` (\*WindowsOptions) - Settings to Windows guest OS customization.
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/CustomizeConfig-not-required.mdx:5]:
```suggestion
- `windows_options` (\*WindowsOptions) - Settings for Windows guest OS customization.
```

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  Specifies whether or not the VM automatically logs on as Administrator.
+
+- `auto_logon_count` (\*int32) - Specifies how many times the VM should auto-logon the Administrator account when auto_logon is true. Default 1
+
+- `admin_password` (\*string) - The new administrator password for this virtual machine.
+
+- `time_zone` (\*int32) - The new time zone for the virtual machine. This is a sysprep-dictated timezone code. Default 85 (GMT)
+
+- `workgroup` (string) - CustomizationIdentification
+  The workgroup for this virtual machine - AD Join is not supported
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:16]:
```suggestion
  The workgroup for the Windows guest operating system.  Joining an Active Directory domain is not supported.
```

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  A list of commands to run at first user logon, after guest customization.
+
+- `auto_logon` (\*bool) - CustomizationGuiUnattended
+  Specifies whether or not the VM automatically logs on as Administrator.
+
+- `auto_logon_count` (\*int32) - Specifies how many times the VM should auto-logon the Administrator account when auto_logon is true. Default 1
+
+- `admin_password` (\*string) - The new administrator password for this virtual machine.
+
+- `time_zone` (\*int32) - The new time zone for the virtual machine. This is a sysprep-dictated timezone code. Default 85 (GMT)
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:13]:
```suggestion
- `time_zone` (\*int32) - The new time zone for the Windows guest operating system. This is a Windows Sysprep timezone code. Default 85 (GMT)
```

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+- `run_once_command_list` (\*[]string) - CustomizationGuiRunOnce
+  A list of commands to run at first user logon, after guest customization.
+
+- `auto_logon` (\*bool) - CustomizationGuiUnattended
+  Specifies whether or not the VM automatically logs on as Administrator.
+
+- `auto_logon_count` (\*int32) - Specifies how many times the VM should auto-logon the Administrator account when auto_logon is true. Default 1
+
+- `admin_password` (\*string) - The new administrator password for this virtual machine.
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:11]:
```suggestion
- `admin_password` (\*string) - The new administrator password for the Windows guest operating system.
```

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 10]
```diff
@@ -0,0 +1,27 @@
+<!-- Code generated from the comments of the WindowsOptions struct in builder/vsphere/clone/step_customize.go; DO NOT EDIT MANUALLY -->
+
+- `run_once_command_list` (\*[]string) - CustomizationGuiRunOnce
+  A list of commands to run at first user logon, after guest customization.
+
+- `auto_logon` (\*bool) - CustomizationGuiUnattended
+  Specifies whether or not the VM automatically logs on as Administrator.
+
+- `auto_logon_count` (\*int32) - Specifies how many times the VM should auto-logon the Administrator account when auto_logon is true. Default 1
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:9]:
```suggestion
- `auto_logon_count` (\*int32) - Specifies how many times the Windows guest operating system should auto-logon the Administrator account when auto_logon is true. Default 1
```

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+- `time_zone` (\*int32) - The new time zone for the virtual machine. This is a sysprep-dictated timezone code. Default 85 (GMT)
+
+- `workgroup` (string) - CustomizationIdentification
+  The workgroup for this virtual machine - AD Join is not supported
+
+- `computer_name` (string) - CustomizationUserData
+  The host name for this virtual machine.
+
+- `full_name` (string) - The full name of the user of this virtual machine. Default: "Administrator"
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:21]:
```suggestion
- `full_name` (string) - The full name of the user of the Windows guest operating system. Default: "Administrator"
```

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+- `admin_password` (\*string) - The new administrator password for this virtual machine.
+
+- `time_zone` (\*int32) - The new time zone for the virtual machine. This is a sysprep-dictated timezone code. Default 85 (GMT)
+
+- `workgroup` (string) - CustomizationIdentification
+  The workgroup for this virtual machine - AD Join is not supported
+
+- `computer_name` (string) - CustomizationUserData
+  The host name for this virtual machine.
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:19]:
```suggestion
  The host name for the Windows guest operating system.
```

File: docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+- `full_name` (string) - The full name of the user of this virtual machine. Default: "Administrator"
+
+- `organization_name` (string) - The organization name this virtual machine is being installed for. Default: "Managed by Packer"
+
+- `product_key` (string) - The product key for this virtual machine.
```
Comment (tenthirtyam) [inline docs-partials/builder/vsphere/clone/WindowsOptions-not-required.mdx:25]:
```suggestion
- `product_key` (string) - The product key for the Windows guest operating system.
```

File: builder/vsphere/clone/step_customize.go [hunk 14]
```diff
@@ -92,37 +121,34 @@ type StepCustomize struct {
...
-		if c.LinuxOptions.Timezone == "" {
-			c.LinuxOptions.Timezone = "UTC"
-		}
+	if options_number > 1 {
+		errs = append(errs, fmt.Errorf("Only one of `linux_options`, `windows_options`, `windows_sysprep_file` can be set"))
+	} else if options_number == 0 {
+		errs = append(errs, fmt.Errorf("One of `linux_options`, `windows_options`, `windows_sysprep_file` must be set"))
 	}
 
-	if len(c.NetworkInterfaces) == 0 {
```
Comment (tenthirtyam) [inline builder/vsphere/clone/step_customize.go:118]: ❓I'm curious why this check for the number of interfaces was removed.

**Review by tenthirtyam — DISMISSED** [review]
Some minor request changes from initial review.

**Comment by tenthirtyam** [comment]
Hi @hmalphettes - have you had an opportunity to review the requested changes?

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by tenthirtyam** [comment]
I’ve committed updates based on my prior suggestions. There’s [one open item](https://github.com/hashicorp/packer-plugin-vsphere/pull/178#discussion_r993648979)  left  for comment., @hmalphettes 

Ryan

File: builder/vsphere/clone/step_customize.go (hunk 14 above)
Comment (nywilken) [reply builder/vsphere/clone/step_customize.go:118]: I do wonder if it was just removed as part of the long if conditional block by accident.

**Comment by nywilken** [comment]
Looks like we need to run `make generate` to update the doc-partials.

**Comment by nywilken** [comment]
Thanks for updating the PR @tenthirtyam. I'm happy to fix the generate check and put back the NetworkInterfaces check, as I it still looks to be needed.

File: builder/vsphere/clone/step_customize.go (hunk 14 above)
Comment (tenthirtyam) [reply builder/vsphere/clone/step_customize.go:118]: I'm thinking so. Should revert that change. I'd be happy to run some tests once I'm back online regularly in about two weeks.

**Comment by tenthirtyam** [comment]
I'm also happy to run some examples (after 11/14) to verify functionality if you're good with that @nywilken.

**Comment by nywilken** [comment]
> I'm also happy to run some examples (after 11/14) to verify functionality if you're good with that @nywilken.

This is good with me @tenthirtyam. Thank you.

File: builder/vsphere/clone/step_customize.go (hunk 14 above)
Comment (tenthirtyam) [reply builder/vsphere/clone/step_customize.go:118]: Thanks for adding it back @nywilken.

Comment (tenthirtyam) [reply builder/vsphere/clone/step_customize.go:118]: It's a requirement to have one network_interface "defined" even if empty.

**Comment by tenthirtyam** [comment]
Local testing shows that these changes perform the intended actions to customize the clone.

However, if the `network_interface` is customized and the IP address changes after the customization Packer will not be able to connect back to the clone to complete the task(s).

Example: DHCP
- DHCP
- Gets 172.16.14.138
- Customized to 172.16.4.200
- Still attempts to connect to 172.16.14.138.

```console
==> vsphere-clone.windows-desktop: Waiting for IP...
==> vsphere-clone.windows-desktop: IP address: 172.16.14.138
==> vsphere-clone.windows-desktop: Using WinRM communicator to connect: 172.16.14.138
==> vsphere-clone.windows-desktop: Waiting for WinRM to become available...
````

File: builder/vsphere/clone/step_customize.go (hunk 14 above)
Comment (nywilken) [reply builder/vsphere/clone/step_customize.go:118]: Thanks for calling this out @tenthirtyam. I saw this requirement in the documentation. Which states "To use DHCP, declare an empty network_interface for each adapter being configured. This field is required". This overrides the network interface settings from the cloned vm correct?

Would it make more sense to not require an empty network configuration and just assume DHCP to start. If a users wishes to customize then the network_interface would not be empty and would work as it does now.

**Comment by nywilken** [comment]
> However, if the network_interface is customized and the IP address changes after the customization Packer will not be able to connect back to the clone to complete the task(s)

By tasks do you mean the provisioning steps that would run on the clone? We have an open issue on Packer about reconnecting after a VM IP change. There is currently no way to reconnect to a source vm if its IP changes. 

If your are referring to other vSphere specific tasks. Is this behavior different from how it works now?

**Comment by tenthirtyam** [comment]
> We have an open issue on Packer about reconnecting after a VM IP change. There is currently no way to reconnect to a source vm if its IP changes.

Yes, this would be the issue here -by defining a customization on the `network_interfaces` the change would not be picked up by Packer; however, I neglected to note that this **is not** a result of this enhancement and it previously existed in the builder.

File: builder/vsphere/clone/step_customize.go (hunk 14 above)
Comment (tenthirtyam) [reply builder/vsphere/clone/step_customize.go:118]: I do think that would be a simpler approach, Wilken.

**Comment by nywilken** [comment]
@tenthirtyam apologies for the delayed response here. Thank you for pushing this forward. At this point I believe this is good to go. The open item about the network interfaces can be a fast follow, as it is change a bit outside of this one. What do you think?

**Review by nywilken — APPROVED** [review]
The overall change looks good to me. We have an open question about the network_interface attribute. If it is a change that we think is blocking this PR I will hold, otherwise this is good go go. Approving in advance.

**Comment by tenthirtyam** [comment]
> The overall change looks good to me. We have an open question about the network_interface attribute. If it is a change that we think is blocking this PR I will hold, otherwise this is good go go. Approving in advance.

I'm good with the status of this pull request - LGTM! 🚀

**Review by tenthirtyam — APPROVED** [review]
Based on my review and testing, I approve.

**Comment by tenthirtyam** [comment]
Possible regression - #242.

**Pull request merged**